pymongo/results.py

```diff
@@ -133,16 +139,18 @@ def raw_result(self) -> Optional[Mapping[str, Any]]:
     @property
     def matched_count(self) -> int:
         """The number of documents matched for this update."""
-        self._raise_if_unacknowledged("matched_count")
-        if self.upserted_id is not None:
-            return 0
+        if self.acknowledged is not None:
+            self._raise_if_unacknowledged("matched_count")
```

Comment:
It would be cleaner to put the `if self.acknowledged is not None:` inside _raise_if_unacknowledged to avoid duplicating it on every call.